File: bdocgen/src/bdocgen/discovery.clj
(ns bdocgen.discovery
  (:require [clojure.string :as str]))

(def ignored-dir-segments
  #{".git" ".tmp" "target" ".venv" "node_modules" ".clj-kondo" ".idea" ".vscode" "_build"})

(defn normalize-path
  [path]
  (-> path
      (str/replace "\\" "/")
      (str/replace #"^\./" "")))

(defn scope->roots
  [scope]
  (case scope
    :self ["bdocgen/docs"]
    :project ["docs" "bdocgen/docs"]
    []))

(defn resolve-roots
  [scope source-roots]
  (if (seq source-roots)
    (->> source-roots
         (map normalize-path)
         vec)
    (scope->roots scope)))

(defn markdown-path?
  [path]
  (or (str/ends-with? path ".md")
      (str/ends-with? path ".markdown")))

(defn hidden-or-ignored-segment?
  [segment]
  (or (contains? ignored-dir-segments segment)
      (and (not= segment "..")
           (str/starts-with? segment "."))))

(defn ignored-path?
  [path]
  (let [segments (str/split path #"/")]
    (some hidden-or-ignored-segment? (butlast segments))))

(defn within-roots?
  [path roots]
  (some (fn [root]
          (or (= path root)
              (str/starts-with? path (str root "/"))))
        roots))

(defn include-path?
  [roots path]
  (let [normalized (normalize-path path)
        resolved-roots (or roots [])]
    (and (within-roots? normalized resolved-roots)
         (markdown-path? normalized)
         (not (ignored-path? normalized)))))

(defn select-doc-paths
  [scope source-roots candidate-paths]
  (let [roots (resolve-roots scope source-roots)]
  (->> candidate-paths
       (map normalize-path)
       (filter (fn [path] (include-path? roots path)))
       distinct
       sort
       vec)))

(defn build-discovery-plan
  [scope source-roots]
  (let [roots (resolve-roots scope source-roots)]
  {:scope scope
   :roots roots
   :accepted-extensions [".md" ".markdown"]
   :ignored-dir-segments (sort ignored-dir-segments)}))

File: bdocgen/src/bdocgen/discovery.py
"""Pure discovery helpers for BDocGen document selection."""

from __future__ import annotations

from pathlib import PurePosixPath

IGNORED_DIR_SEGMENTS = {
    ".git",
    ".tmp",
    "target",
    ".venv",
    "node_modules",
    ".clj-kondo",
    ".idea",
    ".vscode",
    "_build",
}

DEFAULT_SCOPE_ROOTS = {
    "self": ["bdocgen/docs"],
    "project": ["docs", "bdocgen/docs"],
}


def normalize_path(path: str) -> str:
    normalized = path.replace("\\", "/")
    if normalized.startswith("./"):
        return normalized[2:]
    return normalized


def scope_to_roots(
    scope: str,
    default_scope_roots: dict[str, list[str]] | None = None,
) -> list[str]:
    scope_roots = default_scope_roots or DEFAULT_SCOPE_ROOTS
    roots = scope_roots.get(scope)
    if roots:
        return list(roots)
    return []


def resolve_roots(
    scope: str,
    source_roots: list[str] | None,
    default_scope_roots: dict[str, list[str]] | None = None,
) -> list[str]:
    if source_roots:
        return [normalize_path(path) for path in source_roots]
    return scope_to_roots(scope, default_scope_roots)


def markdown_path(path: str) -> bool:
    return path.endswith(".md") or path.endswith(".markdown")


def _hidden_or_ignored_segment(segment: str) -> bool:
    if segment in IGNORED_DIR_SEGMENTS:
        return True
    return segment != ".." and segment.startswith(".")


def ignored_path(path: str) -> bool:
    segments = PurePosixPath(path).parts
    parent_segments = segments[:-1]
    return any(_hidden_or_ignored_segment(segment) for segment in parent_segments)


def _within_roots(path: str, roots: list[str]) -> bool:
    return any(path == root or path.startswith(f"{root}/") for root in roots)


def include_path(roots: list[str], path: str) -> bool:
    normalized = normalize_path(path)
    return (
        _within_roots(normalized, roots)
        and markdown_path(normalized)
        and not ignored_path(normalized)
    )


def select_doc_paths(
    scope: str,
    source_roots: list[str] | None,
    candidate_paths: list[str],
    default_scope_roots: dict[str, list[str]] | None = None,
) -> list[str]:
    roots = resolve_roots(scope, source_roots, default_scope_roots)
    selected = [
        normalize_path(path) for path in candidate_paths if include_path(roots, path)
    ]
    return sorted(set(selected))


def build_discovery_plan(
    scope: str,
    source_roots: list[str] | None,
    default_scope_roots: dict[str, list[str]] | None = None,
) -> dict:
    roots = resolve_roots(scope, source_roots, default_scope_roots)
    return {
        "scope": scope,
        "roots": roots,
        "accepted_extensions": [".md", ".markdown"],
        "ignored_dir_segments": sorted(IGNORED_DIR_SEGMENTS),
    }
